converter_schwab.py

```diff
@@ -0,0 +1,169 @@
+
+            elif (
+                e["Action"] == "Wire Transfer"
+                and e["Description"] == "Cash Disbursement"
+            ):
+                schwab_wire_events.append(
+                    CurrencyConversionRow.from_schwab_json(e).to_dict()
+                )
+
+            elif e["Action"] == "Tax Withholding" and e["Description"] == "Debit":
+                tmp = TaxWithholdingRow.from_schwab_json(e)
+                if len(schwab_dividend_events) > 0 and isinstance(
+                    schwab_dividend_events[-1], DividendRow
+                ):
+                    div = schwab_dividend_events.pop(-1)
+                    div.tax_withholding = tmp.amount
+                    schwab_dividend_events.append(div.to_dict())
+                else:
+                    schwab_dividend_events.append(tmp)
+
+            else:
+                act = e["Action"]
+                des = e["Description"]
+                sym = e["Symbol"]
+                date = e["Date"]
+                warnings.warn(f"skipping {act} on {date} ({sym}: {des})")
+
+    if len(schwab_rsu_lapse_events) != len(schwab_rsu_deposit_events):
+        raise RuntimeError(
+            f"Number of RSU Lapses {len(schwab_rsu_lapse_events)} does not match number of RSU deposits {len(schwab_rsu_deposit_events)}"
+        )
+    elif len(schwab_rsu_lapse_events) > 0:
+        for key, rsu in schwab_rsu_deposit_events.items():
+            if key in schwab_rsu_lapse_events:
+                rsu_lapse = schwab_rsu_lapse_events[key]
+            else:
+                raise ValueError(
+                    f"RSU Deposit {key} does not have a matching Lapse Event"
+                )
+            # schwab applies splits on historical lapse data but not on deposits
+            # thus, use this difference to determine split factor on-the-fly
+            # based on the split factor, we then can rely on the gross quantity
+            # in the lapse event while the prices for the deposit event are already
+            # correct
+            split_factor = rsu_lapse.net_quantity / rsu.net_quantity
+            rsu.gross_quantity = rsu_lapse.gross_quantity / split_factor
+            schwab_rsu_events.append(rsu)
+
+    if len(schwab_espp_events) == 0:
+        schwab_espp_events.append(ESPPRow.empty_dict())
+    if len(schwab_dividend_events) == 0:
+        schwab_dividend_events.append(DividendRow.empty_dict())
+    if len(schwab_sell_events) == 0:
+        schwab_sell_events.append(SellOrderRow.empty_dict())
+    if len(schwab_wire_events) == 0:
+        schwab_wire_events.append(CurrencyConversionRow().empty_dict())
+
+    dfs = {
+        "rsu": pd.DataFrame(schwab_rsu_events),
+        "espp": pd.DataFrame(schwab_espp_events),
+        "dividends": pd.DataFrame(schwab_dividend_events),
+        "buy_orders": pd.DataFrame(schwab_buy_events),
+        "sell_orders": pd.DataFrame(schwab_sell_events),
+        "currency_conversions": pd.DataFrame(schwab_wire_events),
```

Comment:
A wire event doesn't necessarily mean currency conversion. Many people just wire their USD and convert it later (or not at all). But in either case the date will be wrong (which implies wrong exchange rates, too).

The easiest fix would be to leave it on "USD" to "USD" which doesn't imply conversion. But ask those people who _do_ convert their USD to EUR within Schwab, how do their export files look like.

Reply:
An additional thought: I wouldn't put non-converting wire transfer into forex at all. When you sell the stocks, that adds to the balance, but it's already in the "deposits" sheet. Just by transferring money between A and B should not have an effect on your balances. As soon as you do a _real_ conversion (e.g. within Wise or IBKR), that has to appear in the forex sheet separately, but you can't deduct its values from this CSV.

If fees are a concern, then you can add a 0 to 0 USD conversion with a fee of 15 USD.

There are cases when people convert within Schwab, that can go here.

Reply:
It is a good point, but this might require some re-design to begin with.

Technically, for deposits/transfers you nevertheless would have to keep track of buy dates etc. which might complicate things a bit.

Somewhat similar to deposits of shares, ideally, one would have a 
- an extra sheet for transfers of shares between accounts
- an extra sheet for transfers of forex between accounts
One could even add an "account" column to all report values to directly keep track of different things in different accounts to directly have these movements visible. This would complicate the automated reports, however.

I don't really know to be honest.

If people would move some amounts out of Schwab and then do the exchange on there, you nevertheless have to keep track of these movements for FIFO reasons ^^

Reply:
closing as there now is a dedicated movements tab